okhttp/src/main/java/io/grpc/transport/okhttp/OkHttpClientTransport.java

```diff
@@ -319,36 +323,70 @@ private boolean startPendingStreams() {
   @Override
   public void start(Listener listener) {
     this.listener = Preconditions.checkNotNull(listener, "listener");
-    // We set host to null for test.
-    if (host != null) {
-      BufferedSource source;
-      BufferedSink sink;
-      try {
-        socket = new Socket(host, port);
-        if (sslSocketFactory != null) {
-          socket = OkHttpTlsUpgrader.upgrade(
-              sslSocketFactory, socket, authorityHost, port, connectionSpec);
+
+    frameWriter = new AsyncFrameWriter(this, serializingExecutor);
+    outboundFlow = new OutboundFlowController(this, frameWriter);
+
+    // Connecting in the serializingExecutor, so that some stream operations like synStream
+    // will be executed after connected.
+    serializingExecutor.execute(new Runnable() {
+      @Override
+      public void run() {
+        if (isForTest()) {
+          clientFrameHandler = new ClientFrameHandler();
+          executor.execute(clientFrameHandler);
+          connectedCallback.run();
+          frameWriter.setFrameWriter(testFrameWriter);
+          return;
+        }
+        BufferedSource source;
+        BufferedSink sink;
+        try {
+          socket = new Socket(host, port);
+          if (sslSocketFactory != null) {
+            socket = OkHttpTlsUpgrader.upgrade(
+                sslSocketFactory, socket, authorityHost, port, connectionSpec);
+          }
+          socket.setTcpNoDelay(true);
+          source = Okio.buffer(Okio.source(socket));
+          sink = Okio.buffer(Okio.sink(socket));
+        } catch (IOException e) {
+          onIoException(e);
+          // (and probably do all of this work asynchronously instead of in calling thread)
+          throw new RuntimeException(e);
+        }
+
+        FrameWriter rawFrameWriter;
+        synchronized (lock) {
+          if (stopped) {
+            // In case user called shutdown() during the connecting.
+            try {
+              socket.close();
+            } catch (IOException ignore) {
+              // Make checkStyle happy.
+            }
+            return;
+          }
+          Variant variant = new Http2();
+          frameReader = variant.newReader(source, true);
+          rawFrameWriter = variant.newWriter(sink, true);
+          frameWriter.setFrameWriter(rawFrameWriter);
+          clientFrameHandler = new ClientFrameHandler();
```

Comment:
Could you pass frameReader to ClientFrameHandler's constructor? Then in `stopIfNecessary()` decide to do `socket.close();` based on whether `socket` is `null` (see other comment about moving socket assignment into the synchronized block). Then rename the `frameReader` field to `testFrameReader`.